src/s2_cells/mod.rs

```diff
@@ -1,221 +1,263 @@
+        let mut meta = proto::S2Cell::new();
+        meta.set_id(cell_id);
+        meta.set_num_points(self.num_points);
+        meta
+    }
+}
+
+pub struct S2Meta {
+    pub cells: FnvHashMap<CellID, S2CellMeta>,
+    pub attributes: HashMap<String, AttributeDataType>,
+}
+
+impl S2Meta {
+    pub fn to_proto(&self) -> proto::Meta {
+        let cell_protos = self
+            .cells
+            .iter()
+            .map(|(cell_id, cell_meta)| cell_meta.to_proto(cell_id.0))
+            .collect();
+        let mut meta = proto::Meta::new();
+        meta.set_version(CURRENT_VERSION);
+        let mut s2_meta = proto::S2Meta::new();
+        s2_meta.set_cells(::protobuf::RepeatedField::<proto::S2Cell>::from_vec(
+            cell_protos,
+        ));
+        let attributes_meta = self
+            .attributes
+            .iter()
+            .map(|(name, attribute)| {
+                let mut attr_meta = proto::Attribute::new();
+                attr_meta.set_name(name.to_string());
+                attr_meta.set_data_type(attribute.to_proto());
+                attr_meta
+            })
+            .collect();
+        s2_meta.set_attributes(::protobuf::RepeatedField::<proto::Attribute>::from_vec(
+            attributes_meta,
+        ));
+        meta.set_s2(s2_meta);
+        meta
+    }
+
+    pub fn from_proto(meta_proto: proto::Meta) -> Result<Self> {
+        // check if the meta is meant to be for S2 point cloud
+        if !(meta_proto.version == CURRENT_VERSION && meta_proto.has_s2()) {
+            return Err(ErrorKind::InvalidInput("No s2 meta found".to_string()).into());
+        }
+        let s2_meta_proto = meta_proto.get_s2();
+        // cells, num_points
+        let mut cells = FnvHashMap::default();
+        s2_meta_proto.get_cells().iter().for_each(|cell| {
+            let cell_id = CellID(cell.id as u64);
+            cells.insert(
+                cell_id,
+                S2CellMeta {
+                    num_points: cell.num_points,
+                },
+            );
+        });
+
+        // attributes
+        let mut attributes = HashMap::default();
+        // s2_meta_proto.attributes.iter().try_for_each( |attr: proto::Attribute| -> Result<()> {
+        //     let d_type: proto::AttributeDataType = attr.get_data_type();
+        //     let attr_type: crate::AttributeDataType = crate::AttributeDataType::from_proto(d_type)?;//
+        //     attributes.insert(
+        //         attr.name,
+        //         attr_type,
+        //     )
+        // })?;
+        for attr in s2_meta_proto.attributes.iter() {
+            let d_type: proto::AttributeDataType = attr.get_data_type();
+            let attr_type: crate::AttributeDataType = crate::AttributeDataType::from_proto(d_type)?; //
+            attributes.insert(attr.name.to_owned(), attr_type);
+        }
+
+        Ok(S2Meta { cells, attributes })
+    }
+
+    pub fn from_data_provider(data_provider: Box<dyn DataProvider>) -> Result<Self> {

```

Comment:
I think we shouldn't take `data_provider` by value because then you can't use it anywhere else after. How about omething like `from_data_provider(data_provider: &dyn DataProvider)`? You can still use that if the caller has a Box.

Reply:
I've just copied the signature in octree/mod.rs, do you think that it is worth to change the signature also there (in a future pr?)

Reply:
I think it's fine there, because the DataProvider gets moved into the octree, but we just need a reference.